pkg/ccl/streamingccl/logical/logical_replication_writer_processor.go

```diff
@@ -604,39 +611,72 @@ func (lrw *logicalReplicationWriterProcessor) flushBuffer(
 		return b.checkpoint, nil
 	}
 
+	kvs := b.buffer.curKVBatch
+
+	batchSize := int(flushBatchSize.Get(&lrw.EvalCtx.Settings.SV))
+
 	// Ensure the batcher is always reset, even on early error returns.
 	preFlushTime := timeutil.Now()
-	lrw.debug.RecordFlushStart(preFlushTime, int64(len(b.buffer.curKVBatch)))
+	lrw.debug.RecordFlushStart(preFlushTime, int64(len(kvs)))
 
 	// TODO: The batching here in production would need to be much
 	// smarter. Namely, we don't want to include updates to the
 	// same key in the same batch. Also, it's possible batching
 	// will make things much worse in practice.
-	flushByteSize := 0
-	batchStart := 0
-	batchSize := int(flushBatchSize.Get(&lrw.EvalCtx.Settings.SV))
-	batchEnd := min(batchStart+batchSize, len(b.buffer.curKVBatch))
-	for batchStart < len(b.buffer.curKVBatch) && batchEnd != 0 {
-		preBatchTime := timeutil.Now()
-		batchStats, err := lrw.bh.HandleBatch(ctx, b.buffer.curKVBatch[batchStart:batchEnd])
-		if err != nil {
-			// TODO(ssd): Handle errors. We should perhaps split the batch and retry a portion of the batch.
-			// If that fails, send the failed application to the dead-letter-queue.
-			return nil, err
+
+	sort.Stable(roachpb.KeyValueByKey(kvs))
+	var flushByteSize atomic.Int64
+
+	chunkStart, chunkSize := 0, max((len(kvs)/len(lrw.bh))+1, batchSize)
+
+	g := ctxgroup.WithContext(ctx)
+	for worker := range lrw.bh {
+		if chunkStart >= len(kvs) {
+			break
 		}
+		bh := lrw.bh[worker]
+		batchStart := chunkStart
 
-		flushByteSize += batchStats.byteSize
-		batchStart = batchEnd
+		// The chunk should end after the first new key after chunk size.
+		chunkEnd := min(chunkStart+chunkSize, len(kvs))
+		for chunkEnd < len(kvs) && kvs[chunkEnd-1].Key.Equal(kvs[chunkEnd].Key) {
```

Comment:
i don't quite follow this logic. what if `kvs[chunkEnd].Key.Equal(kvs[chunkEnd+1].Key)` is also true?